cmd/state-svc/manager.go

```diff
@@ -5,65 +5,84 @@ import (
 	"io/ioutil"
 	"net/http"
 	"os"
-	"path/filepath"
 	"time"
 
 	"github.com/shirou/gopsutil/process"
+	"github.com/spf13/cast"
 
 	"github.com/ActiveState/cli/cmd/state-svc/internal/server"
 	"github.com/ActiveState/cli/internal/config"
 	"github.com/ActiveState/cli/internal/constants"
 	"github.com/ActiveState/cli/internal/errs"
 	"github.com/ActiveState/cli/internal/exeutils"
 	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/svcmanager"
 )
 
 type serviceManager struct {
-	cfg    *config.Instance
-	lockFp string
+	cfg *config.Instance
 }
 
 func NewServiceManager(cfg *config.Instance) *serviceManager {
-	return &serviceManager{cfg, filepath.Join(cfg.ConfigPath(), "state-svc.lock")}
+	return &serviceManager{cfg}
 }
 
 func (s *serviceManager) Start(args ...string) error {
-	curPid, err := s.Pid()
-	if err == nil && curPid != nil {
-		return errs.New("Service is already running")
-	}
+	var proc *os.Process
+	err := s.cfg.SetWithLock(constants.SvcConfigPid, func(oldPidI interface{}) (interface{}, error) {
+		oldPid := cast.ToInt(oldPidI)
+		curPid, err := s.CheckPid(oldPid)
+		if err == nil && curPid != nil {
+			return nil, errs.New("Service is already running")
+		}
 
-	proc, err := exeutils.ExecuteAndForget(args[0], args[1:])
-	if err != nil {
-		return errs.New("Could not start serviceManager")
-	}
+		proc, err = exeutils.ExecuteAndForget(args[0], args[1:])
+		if err != nil {
+			return nil, errs.New("Could not start serviceManager")
+		}
 
-	if proc == nil {
-		return errs.New("Could not obtain process information after starting serviceManager")
-	}
+		if proc == nil {
+			return nil, errs.New("Could not obtain process information after starting serviceManager")
+		}
 
-	if err := s.cfg.Set(constants.SvcConfigPid, proc.Pid); err != nil {
-		if err := proc.Signal(os.Interrupt); err != nil {
-			logging.Errorf("Could not cleanup process: %v", err)
-			fmt.Printf("Failed to cleanup serviceManager after lock failed, please manually kill the following pid: %d\n", proc.Pid)
+		return proc.Pid, nil
+	})
+	if err != nil {
+		if proc != nil {
+			if err := proc.Signal(os.Interrupt); err != nil {
+				logging.Errorf("Could not cleanup process: %v", err)
+				fmt.Printf("Failed to cleanup serviceManager after lock failed, please manually kill the following pid: %d\n", proc.Pid)
+			}
 		}
 		return errs.Wrap(err, "Could not store pid")
 	}
 
 	logging.Debug("Process started using pid %d", proc.Pid)
-
 	return nil
 }
 
 func (s *serviceManager) Stop() error {
-	pid, err := s.Pid()
+	err := s.cfg.Reload()
+	if err != nil {
+		return errs.Wrap(err, "Failed to reload configuration")
+	}
+	pid, err := s.CheckPid(s.cfg.GetInt(constants.SvcConfigPid))
 	if err != nil {
 		return errs.Wrap(err, "Could not get pid")
 	}
 	if pid == nil {
 		return nil
 	}
 
+	waitMgr := svcmanager.New(s.cfg)
+	err = waitMgr.Wait()
+	if err != nil {
```

Comment:
Why is this necessary? We already checked the PID.

Additionally this is assuming the meaning of `err`.

Reply:
I am pretty sure that this happened in one of our integration test runs at some point.

The PID value had already been written, but the server has not been started yet ie., it has not written the PORT value to the configuration file. So, the `waitMgr` just waits for the process to start up such that it can talk to it via the port...

I am fine with removing it for now, and we should just be vigilant and see if it occurs again.  It may not be a real issue...

Reply:
I am pretty sure that this happened in one of our integration test runs at some point.

The PID value had already been written, but the server has not been started yet ie., it has not written the PORT value to the configuration file. So, the `waitMgr` just waits for the process to start up such that it can talk to it via the port...

I am fine with removing it for now, and we should just be vigilant and see if it occurs again.  It may not be a real issue...

Reply:
This is on "Stop" though?

Reply:
Yes, it is kind of silly to wait for the service to start in order to stop it.  But I saw at least one instance where it tried to do that on port `0`.  And that indicates that the port number was not written to the configuration yet...

Reply:
Let's remove this. It sounds to me like this is here to work around an issue we don't fully understand, I'd rather we understand the issue and solve it properly. I can't imagine a legitimate use-case for this.